llvm/lib/CAS/CachingOnDiskFileSystem.cpp

```diff
@@ -64,17 +64,28 @@ class CachingOnDiskFileSystemImpl final : public CachingOnDiskFileSystem {
 
   Expected<DirectoryEntry *> makeSymlink(DirectoryEntry &Parent,
                                          const std::string &TreePath);
+  Expected<DirectoryEntry *>
+  makeSymlinkTo(DirectoryEntry &Parent, StringRef TreePath, StringRef Target);
 
   Expected<DirectoryEntry *> makeFile(DirectoryEntry &Parent,
                                       StringRef TreePath, int BorrowedFD,
                                       sys::fs::file_status Status);
 
   /// Preload the real path for \p Remaining, relative to \p From.
-  Error preloadRealPath(DirectoryEntry &From, StringRef Remaining);
+  /// \returns The real path entry if it can be computed, an error if the path
+  ///          cannot be accessed, or nullptr if the path was accessed but there
+  ///          was a subsequent filesystem modification.
+  Expected<FileSystemCache::DirectoryEntry *>
+  preloadRealPath(DirectoryEntry &From, StringRef Remaining);
 
   /// Look up a name on disk inside \p From.
   Expected<DirectoryEntry *> lookupOnDiskFrom(DirectoryEntry &Parent,
-                                              StringRef Name);
+                                              StringRef Name,
+                                              DirectoryEntry *KnownRealPath);
+
+  Expected<DirectoryEntry *>
+  lookupOnDiskFrom(DirectoryEntry &Parent, const std::string &TreePath,
```

Comment:
Why is this not a StringRef?

Reply:
Also, wondering if the difference between the name and treepath parameters above is a bit subtle for an overload. Maybe these should have different names?

Reply:
> Why is this not a StringRef?

It calls `makeSymlink`, which takes a `const std::string &` in order to use `c_str()` without a copy. Since it's already a string in the callers, I used the same pattern here.  Could use StringRef and copy it; I guess that's what it would do if it were using a normal filesystem API that takes a Twine anyway, this is just an artifact of the fact we're calling readlink directly.

> Maybe these should have different names?

I considered `makeEntryFromStatus` but that's only really accurate when you pass a `KnownStatus`.  Maybe `lookupTreePathOnDisk`?

Reply:
> which takes a const std::string & in order to use c_str() without a copy

Aha; that's subtle. Maybe worth a comment?

(Really we should have a `NullTerminatedStringRef : public StringRef` whose type remembers that it's null-terminated, which we can use in situations like this. I've looked at it... a bit hard, but probably worth it.)

Reply:
Or I see you committed; seems fine to have the copy. We can fix later if it matters, which it probably doesn't.

Reply:
> I considered makeEntryFromStatus but that's only really accurate when you pass a KnownStatus. Maybe lookupTreePathOnDisk?

What you landed looks great.